Deprecate `$` as bitwise xor, replace by xor(), as discussed in #18696

davidavdav opened the pull request:
These two commits implement the replacement of `$` (bitwise exclusive or) by `xor()`.  The first commit includes all changes that are enough to pass the Julia unit tests.  The second commit implements more changes in Base, found by grepping for `' $ '`.  

There still are 3851 lines in base with a `$`, but it is not directly clear if any of them is a `xor` or not (pun intended).  This would require a Julia parser perhaps.   It would be rather handy to be able to find all usages of `$` as xor in the code.

davidavdav committed:
Deprecate $ for xor() (#18696)

This includes:
 - occurrances of $ (as infix operator) and $= in base
 - occurances of $ in test
 - the definition in the manual
 - the entry in REPL help

This commit does not introduce an alternative,
that will happen in the next commit.

vchuravy commented:
I am not a big fan of this change. The shorthand form is really useful for implementing algorithm that require bit-twiddling . I see where the need for deprecation is coming from, but could we atleast then also add a Unicode infix alias? \veebar or \oplus should be the right ones.

davidavdav commented:
Sure, we can add one of these as a replacement, why not?  This PR make searching for xor-usage of `$` a little bit easier, although there is no guarantee it is complete.

davidavdav commented on base/sparse/sparsematrix.jl:
```diff
@@ -904,7 +904,8 @@ function _ispermutationvalid_permute!{Ti<:Integer,Tp<:Integer}(perm::AbstractVec
     checkspace[1:n] = 0
     for k in perm
-        (0 < k <= n) && ((checkspace[k] $= 1) == 1) || return false
+        checkspace[k] = xor(checkspace[k], 1)
```
Well spotted.   I didn't think about the shortcut behaviour.

carlobaldassi commented:
> could we atleast then also add a Unicode infix alias? \veebar or \oplus should be the right ones.

\veebar seems the correct one. At which point, we might as well have \vee, \wedge etc.

Though it may be worth noting that although having an infix substitute goes a long way, full "feature parity" will only be achieved when #15964 is fixed (cf. [this](https://github.com/JuliaLang/julia/pull/18977/files#diff-f8884151cebf8b66d7dba5eb76b46128L118) and following).

davidavdav commented:
It looks like `⊻` already is an operator. (Where is this defined?)  
```julia
⊻ = xor
```
should be enough, then?

But anyhow, this PR is not really about the addition of alternative operators.  There is some discussion about that in #18696 and references therein.  It looks like once there is consensus about the unicode operators, it is easy enough to add them.  (And if people feel that should happen in this PR, I am happy to add them.)  

Deprecation of `$` (with the purpose of freeing it for some syntactic function in the future) needs to happen first, I believe.  I've just gone through all `$`s in base, and at the moment it already appears to have quite varied semantics.

tkelman commented:
> It looks like `⊻` already is an operator. (Where is this defined?) 

https://github.com/JuliaLang/julia/blob/5722529ac376012f845b8d12cfabff361148c5f4/src/julia-parser.scm#L24

nalimilan commented:
> Deprecation of $ (with the purpose of freeing it for some syntactic function in the future) needs to happen first, I believe. I've just gone through all $s in base, and at the moment it already appears to have quite varied semantics.

I disagree. It would be much easier to just replace `$` with `⊻`, and mention this in the deprecation warning, than to remove all uses of `$` first in Base and in packages, and only then introduce `⊻`.

KristofferC commented:
You need `const` if you want to do the definition as above.

simonbyrne commented:
It might also be worth adding `\xor` tab completion (I'm never going to remember `\veebar`)

davidavdav commented:
> It might also be worth adding `\xor` tab completion (I'm never going to remember `\veebar`)

That is a good idea.

JeffBezanson commented:
> I've just gone through all $s in base, and at the moment it already appears to have quite varied semantics.

I'm aware of 2; xor (infix) and interpolation (prefix). Are there more?

davidavdav commented:
> I'm aware of 2; xor (infix) and interpolation (prefix). Are there more?

I won't doubt you on Julia parsing or semantics :-)

Just grepping for `$` I see a lot of `$(...)` constructions which probably are just interpolation, there is the odd end-of-string match re, sometimes the operator itself (as in `export $`)---I was just looking for the infix meaning.  I hope to have found them all.

vchuravy commented on base/associative.jl:
```diff
@@ -252,7 +252,7 @@ const hasha_seed = UInt === UInt64 ? 0x6d35bb51952d5539 : 0x952d5539
     h = hash(hasha_seed, h)
     for (k,v) in a
-        h $= hash(k, hash(v))
+        h = h ⊻ hash(k, hash(v))
```
You could add `⊻=` in https://github.com/JuliaLang/julia/blob/5722529ac376012f845b8d12cfabff361148c5f4/src/julia-parser.scm#L11

davidavdav replied:
OK, sure, thanks, but is this something for #15964, or is that about a more general approach?

Should I also deprecate `$=` there then, to really free `$` for future syntactical purposes?

vchuravy reviewed (commented):
If we are using `\veebar` we should maybe remove the `.\veebar` version? 
https://github.com/JuliaLang/julia/blob/5722529ac376012f845b8d12cfabff361148c5f4/src/julia-parser.scm#L23-L24

simonbyrne commented:
I don't think we should include ∧ and ∨ in this PR: see #17472 for some discussion of this issue.

simonbyrne commented:
Also, I think we should keep `\veebar`, there's no problem with having 2 aliases

davidavdav committed:
Add ⊻ as alternative to xor, and `⊻=` as alternative to `$=`.

Please note that `⊻=` is in the add-dots group in the parser, preparing for `.⊻=`.
However, this commit does not implement `.⊻` yet, this should happen in
a separate commit with `.&` and `.|` together.

davidavdav commented:
All right, I rebased everything a few times---things should be in line with master-as-we-write.  I kept three commits, so there is a choice what to include (either one, two, or all three commits)

- 63bbaef, deprecates `$` and replaces infix-`$` occurrences in base with `xor()`
- e296b3e, introduces `⊻` and `⊻=`, and replaces most of the above `xor()`s in base with these.  The result of these two commits compares most favourably with master, as it includes mainly `$`⟶`⊻` diffs. 
- e4a340c, also introduces `∨` and `∧` as synonyms to `|` and `&`, as some people in this discussion suggested.  This change is still under debate, I gather, so this commit can easily be cherry-unpicked.  

I hope this rebase satisfies everyone who have taken part in this discussion.

tkelman commented:
please separate the third commit to a different pull request

davidavdav commented:
> please separate the third commit to a different pull request

No problem, I removed the third commit from this PR and updated the comment above accordingly.

tkelman commented on base/show.jl:
```diff
@@ -449,7 +449,7 @@ const uni_ops = Set{Symbol}([:(+), :(-), :(!), :(¬), :(~), :(<:), :(>:), :(√)
     :(=), :(+=), :(-=), :(*=), :(/=), :(\=), :(^=), :(&=), :(|=), :(÷=), :(%=), :(>>>=), :(>>=), :(<<=),
     :(.=), :(.+=), :(.-=), :(.*=), :(./=), :(.\=), :(.^=), :(.&=), :(.|=), :(.÷=), :(.%=), :(.>>>=), :(.>>=), :(.<<=),
-    :(&&), :(||), :(<:), :(=>), :($=)])
+    :(&&), :(||), :(<:), :(=>), :($=), :(⊻=)])
```
guess `$=` should be marked somehow as "remove from this list when deprecation is removed," right?

davidavdav replied:
Yes, you are right there.  I left `$=` in there because it may still be used and it should be shown correctly in expressions.

simonbyrne replied:
Can you add a comment as Tony mentioned? Something like
```
# $= was previously used for xor, issue #18977
```

tkelman commented on src/julia-parser.scm:
```diff
@@ -8,7 +8,7 @@
 (define prec-assignment
   (append! (add-dots '(= += -= *= /= //= |\\=| ^= ÷= %= <<= >>= >>>= |\|=| &=))
-           '(:= => ~ $=)))
+           '(:= => ~ $= ⊻=)))
```
`⊻=` can be under add-dots now right? I think `$=` wasn't added there mostly because `$` has other meanings that were making it complicated to parse correctly

davidavdav replied:
I am not sure---I have to admit that my lisp is not up to scratch.  From the code I understand that the `add-dots` operators add a dotted version as well.  `⊻=` already works element wise.  But there is no broadcasting `.⊻` indeed.  Should that be added?

davidavdav commented on src/julia-parser.scm:
```diff
@@ -78,7 +78,7 @@
 (define syntactic-operators
   (append! (add-dots '(= += -= *= /= //= |\\=| ^= ÷= %= <<= >>= >>>= |\|=| &=))
-           '(:= --> $= => && |\|\|| |.| ... ->)))
+           '(:= --> $= ⊻= => && |\|\|| |.| ... ->)))
```
@tkelman I suppose this `⊻=` should then also move one line up to `add-dots`, right?

davidavdav commented on src/julia-syntax.scm:
```diff
@@ -2060,6 +2060,7 @@
    '&=     lower-update-op
    '.&=     lower-update-op
    '$=     lower-update-op
+   '⊻=     lower-update-op
```
@tkelman ...and here I should then add a line with '.⊻= ...

davidavdav commented:
Weekly rebase, to keep up with master. 

- https://github.com/JuliaLang/julia/pull/18977/commits/5e853bb5acd471d638e58432b658b437b20dd6d5 deprecates `$` and replaces infix-`$` occurrences in base with `xor()`
- https://github.com/JuliaLang/julia/pull/18977/commits/ce94dff4e33dbc3c19539a424f6a706cd7d39435 introduces `⊻` and `⊻=`, and replaces most of the above `xor()`s in base with these. The result of these two commits compares most favourably with master, as it includes mainly `$⟶⊻` diffs.

This rebase includes the latest changes suggested in a review by @tkelman. 

Incidentally, I am planning another PR that will introduce `.&`, `.|` and `.⊻` broadcasting operators, which will automatically enable the use of `.&=`, `.|=` and `.⊻=`.  

`xor.()` already automagically works in this commit thanks to #15032.

davidavdav commented:
Weekly rebase, to keep up with master.

simonbyrne commented:
Can you add an item to `NEWS.md`?

davidavdav committed:
Add deprecation comment in show.jl and deprecation item in NEWS.md

simonbyrne commented:
Can you resolve the conflict with NEWS.md? Then I think this is ready to go.

davidavdav commented:
The build failure on linux is not related to this PR, I think.

musm commented:
although I'm against this change, it seems strange to just deprecate `$`, while still keeping `&` and `|`. If this change passes, It would make sense to change all three for consistency, otherwise it's confusing that `&` behaves as you expect,  but   `$` doesn't.

simonbyrne commented:
Done! Thanks for your patience, modifying that many files in one go is always a bit tricky.

simonbyrne commented:
@musm The distinction is that `&` and `|` have a long history as bitwise and/or (dating back to C at least). The use of `$` for `xor` on the other hand was very much nonstandard (C uses `^`, which we already use for raising to a power), and so I don't think changing this affect anyone's expectations.

musm commented:
fair point. Doesn't someone have to update the syntax highlighting on github etc. for this change? It used to be that `$` got highlighted.

mgkuhn commented:
In the computing literature, I encounter ⊕ (U+2295 CIRCLED PLUS) far more commonly as the XOR operator [same as addition over GF(2^n)] than ⊻ (U+22BB). How about supporting both?

nalimilan commented:
In general it's not a good idea to support multiple ways of doing the same thing in a language (an exception to this is ASCII vs. Unicode operators for practical reasons, but at least this is a consistent rule). You can define it for yourself if you really find `⊕` better.

tkelman commented:
U+22BB is called xor in unicode. I see ⊕ used for dilation/Minkowski sum pretty often, it's context-dependent and a useful operator to leave available since it doesn't have a single unambiguous interpretation.

carlobaldassi commented:
Besides what @nalimilan and @tkelman wrote, I would guess that the reason why ⊕ is commonly used for xor in the literature is that for 0/1 variables xor is the same as "addition modulo 2". However, when xor'ing 2 integers you expect a bitwise xor, so ⊕ doesn't seem right to me in that context.

mgkuhn commented:
@tkelman I would only suggest to define ⊕ as carry-less addition (aka XOR) for Unsigned or Integer, as the operator makes no sense for other types (e.g. floating point). That excludes conflicts with non-integer uses of the same symbol, e.g. the ones in relativity you mention.

@carlobaldassi If you cut off the carry-over bits in n-bit modular integer arithmetic, the resulting operations are called binary Galois Field arithmetic, often abbreviated as GF(2^n). The GF(2^n) addition is exactly the n-bit XOR over unsigned integers. Modulo 2 arithmetic is just a special case n=1 of Galois Field arithmetic, commonly labeled [GF(2)](https://en.wikipedia.org/wiki/GF(2)).

Galois Field arithmetic is widely used in cryptography and in error-correction algorithms, as it has similar algebraic properties as integers, without the hardware cost of having to wait for carry-over ripple in your adder. Intel and AMD CPUs since 2010 have included a [carry-less multiplication](https://en.wikipedia.org/wiki/CLMUL_instruction_set) instruction for it.

I would suggest to define in Julia the operators ⊕ and ⊗ for carry-less addition (aka XOR) and carry-less multiplication (CLMUL) between unsigned integers.

carlobaldassi commented:
@mgkuhn that was exactly my point: addition on GF(n) and xor are only equivalent for n=2. When you do e.g. `xor(a::UInt, b::UInt)` you expect bitwise xor instead, treating the UInts as bit masks. So ⊕ is not suited for that. For operations on GF(n), you'd better implement your own integer type with `n` as a type parameter, at which point you can just overload `+` and `*`.

StefanKarpinski commented:
There are also many other possible meanings of ⊕ and ⊗ in other areas (e.g. ⊗ as a tensor product). You're free to define these operators in your own projects, but it's inappropriate for the base language to force specific meanings on symbols that have multiple standard meanings.
